playground/serve.go

```diff
@@ -16,27 +16,243 @@
+		return
+	}
+	components := getComponentsAndUpdateIfStale(r)
+	if components == nil {
+		http.Error(w, "Server error", http.StatusServiceUnavailable)
+		return
+	}
+
+	w.Header().Set("Content-type", "application/json")
+	w.Write(components.Components)
+}
+
+func getComponentsAndUpdateIfStale(r *http.Request) *AmpComponentsList {
+	curTime := int(time.Now().Unix())
+	ctx := appengine.NewContext(r)
+	latest, _ := fetchComponentsFromMemCache(ctx)
+	// latest could be nil if not in memcache or datastore.
+	var timestamp int
+	if latest != nil {
+		timestamp = latest.Timestamp
+	}
+
+	if curTime-timestamp > COMPONENTS_UPDATE_FREQ_SECONDS {
+		createTaskQueueUpdate(ctx, timestamp)
+	}
+	return latest
+}
+
+func createTaskQueueUpdate(ctx context.Context, timestamp int) {
+	t := taskqueue.NewPOSTTask(PLAYGROUND_PATH_PREFIX+"/amp-component-versions-task",
+		url.Values{})
+	// Setting the name explicitly means that only one task will ever
+	// be in the queue, even if attempted by separate instances.
+	// The name chosen includes the timestamp of the last known
+	// components list (or 0 for the first attempt), which means
+	// that only one update task can be created for each version
+	// known.
+	t.Name = fmt.Sprintf("amp-components-list-last-known-%d", timestamp)
+	minBackoff, _ := time.ParseDuration("10s")
+	t.RetryOptions = &taskqueue.RetryOptions{
+		RetryLimit: 5,
+		MinBackoff: minBackoff,
+	}
+	log.Infof(ctx, "Adding to taskqueue: %s", t.Name)
+	taskqueue.Add(ctx, t, "")
+}
+
+func componentsTask(w http.ResponseWriter, r *http.Request) {
+	ctx := appengine.NewContext(r)
+	_, err := fetchAndUpdateComponents(ctx)
+	if err != nil {
+		http.Error(w, "Server error", http.StatusInternalServerError)
+		return
+	}
+	w.Write([]byte("Updated components"))
+}
+
+func fetchAndUpdateComponents(ctx context.Context) ([]byte, error) {
+	log.Infof(ctx, "Fetching components list from GitHub")
+	components, err := fetchComponents(ctx)
+	if err != nil {
+		return nil, err
+	}
+	addComponentsToStores(ctx, components)
+	return components, nil
+}
+
+func fetchComponentsFromDataStore(ctx context.Context) (*AmpComponentsList, error) {
+	log.Infof(ctx, "Retrieving components from datastore")
+	var components AmpComponentsList
+	key := datastore.NewKey(ctx, "AmpComponentsList", "ComponentsListKey", 0, nil)
+	err := datastore.Get(ctx, key, &components)
+	if err != nil {
+		log.Infof(ctx, "Error retrieving components from datastore")
+		rawComponents, err := fetchAndUpdateComponents(ctx)
+		if err != nil {
+			return nil, err
+		}
+		components.Components = rawComponents
+		components.Timestamp = int(time.Now().Unix())
+	}
+	return &components, err
+}
+
+func fetchComponentsFromMemCache(ctx context.Context) (*AmpComponentsList, error) {
+	log.Infof(ctx, "Retrieving components from memcache")
+	var components AmpComponentsList
+	_, err := memcache.Gob.Get(ctx, COMPONENTS_MEMCACHE_KEY, &components)
+	if err == memcache.ErrCacheMiss {
```

Comment:
When I deploy the version, the result is always `{}`. It's related to this part as it works when I replace this line with ` if err != nil {`. BTW this is only a problem when I deploy it to app engine, it works locally.

In general can you make sure to handle the errors correctly:

* can you log all errors: `log.Errorf(ctx, "..... %v", err)`
* can you make sure to return a corresponding status code if we could not retrieve the list of components.